libimage/runtime.go

```diff
@@ -308,29 +308,53 @@ func (r *Runtime) LookupImage(name string, options *LookupImageOptions) (*Image,
 
 	// The specified name may refer to a short ID. Note that this *must*
 	// happen after the short-name expansion as done above.
-	img, err := r.lookupImageInLocalStorage(name, name, options)
+	img, err := r.lookupImageInLocalStorage(name, name, nil, options)
 	if err != nil {
 		return nil, "", err
 	}
 	if img != nil {
 		return img, name, err
 	}
 
-	return r.lookupImageInDigestsAndRepoTags(name, options)
+	return r.lookupImageInRepoTags(name, possiblyUnqualifiedNamedReference, options)
 }
 
 // lookupImageInLocalStorage looks up the specified candidate for name in the
 // storage and checks whether it's matching the system context.
-func (r *Runtime) lookupImageInLocalStorage(name, candidate string, options *LookupImageOptions) (*Image, error) {
+func (r *Runtime) lookupImageInLocalStorage(name, candidate string, namedCandidate reference.Named, options *LookupImageOptions) (*Image, error) {
 	logrus.Debugf("Trying %q ...", candidate)
-	img, err := r.store.Image(candidate)
-	if err != nil && !errors.Is(err, storage.ErrImageUnknown) {
-		return nil, err
-	}
-	if img == nil {
-		return nil, nil
+
+	var err error
+	var img *storage.Image
+	var ref types.ImageReference
+
+	// for images pulled by tag, Image.Names does not currently contain a
+	// repo@digest value, so such an input would not match.
+	if namedCandidate != nil {
+		ref, err = storageTransport.Transport.NewStoreReference(r.store, namedCandidate, "")
+		if err != nil {
+			return nil, err
+		}
+		img, err = storageTransport.Transport.GetStoreImage(r.store, ref)
+		if err != nil {
+			if errors.Is(err, storage.ErrImageUnknown) {
+				return nil, nil
+			}
+			return nil, err
+		}
+		// NOTE: we must reparse the reference another time below since
+		// an ordinary image may have been referenced via its parent
+		// (manifest list) digest.
```

Comment:
BTW I’m not sure this works — in particular it relies on `image.IsManifestList`, which ultimately (due to `ImageDigestBigDataKey`) checks the _last-written_ manifest. On a multi-arch image pull, that’s actually never going to be the multi-arch manifest AFAICS.

But it’s also, I think, not a new behavior — i.e. I now think I was wrong in https://github.com/containers/common/pull/1505#discussion_r1239074353 .

Unless I’m missing something, this was broken for tags before (in the fairly rare situation where we have two different variants of a single multi-arch image pulled locally); now it is also broken for digests, but in cases where digest lookups didn’t work _at all_, so it’s on net an improvement.

Is that a correct analysis, or am I completely missing something important?

If this does fail as suggested above, at least a `TODO` comment either in the call site of `image.IsManifestList`, or in the implementation, seems worthwhile.

Reply:
I still need to wrap my head around this one.  What I get without reparsing the reference is

> time="2023-06-27T09:49:28+02:00" level=error msg="Looking up docker.io/library/busybox@sha256:6e494387c901caf429c1bf77bd92fb82b33a68c0e19f6d1aa6a3ac8d27a704
9d in local storage: unable to load manifest list: unsupported format \"application/vnd.docker.distribution.manifest.v2+json\": manifest type not supported"
Trying to pull docker.io/library/busybox@sha256:6e494387c901caf429c1bf77bd92fb82b33a68c0e19f6d1aa6a3ac8d27a7049d...

Where `IsManifestList` returns true and we later on fail with the upper error.

Reply:
Added a `// FIXME`

Reply:
> What I get without reparsing the reference is

So the situation is a lookup with an explicit digest reference to the manifest list, and the code in this PR modified to use the `ref, err = storageTransport.Transport.NewStoreReference(r.store, namedCandidate, "")` value only?

In that case (reading the code, not testing it), I think
- `lookupImageInLocalStorage` _with the hypothetical modified version of this PR_ creates a `libimage.Image` with `storageReference` based on `namedCandidate`
  - (i.e. not an immutable reference; I’m not sure that’s wanted)
- c/image, given such a `storageReference` reliably returns the manifest list (it chooses a manifest by the provided digest)
- `libimage.Image.IsManifestList` returns `true`
- but `libimage.Image.ToManifestList` is may fail: that code path discards the users’ digest reference, resolves the image to `storage.Image.ID`, and then `libimage.manifests.LoadFromImage` loads the last-written manifest (which is, per the above, never the multi-arch one).


IOW it seems to me that the previous version of `lookupImageInLocalStorage`, and the version of `lookupImageInLocalStorage` as it is right now in this PR, create a `storageReference` purely based on `Image.ID`; and then the problems of `libimage` and `libimage/manifest` only reading the `ImageDigestBigDataKey` cancel out, because the path that checks for a multi-arch image and the path that loads a multi-arch manifest both see the same manifest. If `lookupImageInLocalStorage` is changed to set a `storageReference` directly to a name@digest, `IsManifestList` and `ToManifestList` may load different manifests and get out of sync.

Reply:
Overall I now fairly strongly suspect that `options.{Architecture,OS,Variant}` very rarely work in `lookupImageInLocalStorage`.

I’m not sure how much that matters — I suppose in the common case only one platform instance of an image exists in any storage, and as long as we _pull_ the right one, we will find the only one pulled and use it. But it there were important Podman workflows where local storage simultaneously contained instances from multiple platforms, I wouldn’t know.

Reply:
I am not entirely sure how we can resolve the issue.  Ultimately, multiple digests may resolve to one image.  If we query storage with the "parent digest" (i.e., of the manifest list), we run into the upper issue and `ImageDigestBigDataKey` returning the other raw manifest.

I have a feeling that for digested inputs, we need to do the repo-digest dance?  That is: find a matching image for the repo first, then check whether it has the requested digest.

Reply:
c/image does even more than that — if there are multiple images with the matching manifest digest (i.e. the digest is of the multi-arch manifest), it chooses the image that matches the requested architecture.

I think that’s certainly possible to invoke in this function (with a c/image extension as discussed in https://github.com/containers/common/pull/1505#discussion_r1239038911 ). I haven’t looked into other parts of the `libimage.Image` and `libimage/manifest` API — if they identify an image by ID, it might be difficult.

Reply:
Do you think https://github.com/containers/common/pull/1505#discussion_r1239038911 is a hard requirement or a Podman 4.7 goal?  If it's hard requirement, then this PR will not make it into the 4.6 release.  I would have loved it to make into the release but prefer doing it right.